crates/hulk_nao/src/camera.rs

```diff
@@ -31,60 +36,73 @@ impl Camera {
         parameters: Parameters,
         i2c_head_mutex: Arc<Mutex<()>>,
     ) -> Result<Self> {
-        let (sender, receiver) = watch::channel(None);
+        let (sender, _) = watch::channel(None);
         let camera = Self {
-            camera: RwLock::new(None),
-            path: path.as_ref().to_path_buf(),
-            camera_position,
-            read_mutex: Mutex::new(()),
-            parameters,
-            i2c_head_mutex,
+            camera: Mutex::new(CameraHardware {
+                i2c_head_mutex,
+                config: CameraConfiguration {
+                    path: path.as_ref().to_path_buf(),
+                    parameters,
+                },
+                camera: None,
+                camera_position,
+            }),
             image_sender: sender,
-            image_receiver: receiver,
         };
-        camera.reset().wrap_err("failed to reset")?;
+
+        camera.camera.lock().reset().wrap_err("failed to reset")?;
+
         Ok(camera)
     }
 
     pub fn read(&self) -> Result<YCbCr422Image> {
-        let mut this_receiver = self.image_receiver.clone();
-        this_receiver.mark_unchanged();
-        if let Some(_lock) = self.read_mutex.try_lock() {
-            self.wait_for_device()
-                .wrap_err("failed to wait for device")?;
+        let mut this_receiver = self.image_sender.subscribe();
 
-            let mut camera_lock = self.camera.write();
-            let camera = camera_lock.as_mut().unwrap();
-            let buffer = camera.dequeue().wrap_err("failed to dequeue buffer")?;
-            camera
-                .queue(vec![
-                    0;
-                    match self.parameters.format {
-                        nao_camera::Format::YUVU =>
-                            (4 * self.parameters.width / 2 * self.parameters.height) as usize,
-                    }
-                ])
-                .wrap_err("failed to queue buffer")?;
-            let new_image = YCbCr422Image::from_raw_buffer(
-                self.parameters.width / 2,
-                self.parameters.height,
-                buffer,
-            );
-            self.image_sender.send(Some(new_image.clone()))?;
+        if let Some(mut camera) = self.camera.try_lock() {
+            let new_image = camera.get_next_image()?;
+            self.image_sender.send(Some(new_image.clone()));
             return Ok(new_image);
         }
-        let image = this_receiver.borrow_and_update().clone();
+
+        let image = this_receiver.wait();
         Ok(image.unwrap())
         // TODO: read consecutive sequence number checking
     }
+}
+
+impl CameraHardware {
+    fn get_next_image(&mut self) -> Result<YCbCr422Image> {
+        self.wait_for_device()
+            .wrap_err("failed to wait for device")?;
 
-    fn wait_for_device(&self) -> Result<()> {
+        let Some(camera) = self.camera.as_mut() else {
+            bail!("camera does not exist")
+        };
```

Comment:
Isn't there a monadic way of expressing this `let Some else`? Like `unwrap_or` or `or_else`. I remember some method for converting a `None` into a `Err(...)`. 🤔

```suggestion
        let camera = self.camera.as_mut().ok_or_else(|| eyre!("camera does not exist"))?;
```